receiver/dockerstatsreceiver/receiver.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dockerstatsreceiver
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumerdata"
+	"go.opentelemetry.io/collector/obsreport"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/redisreceiver/interval"
+)
+
+var _ component.MetricsReceiver = (*Receiver)(nil)
+
+type Receiver struct {
+	config       *Config
+	logger       *zap.Logger
+	nextConsumer consumer.MetricsConsumerOld
+	client       *DockerClient
+	runner       *interval.Runner
+	obsCtx       context.Context
+	runnerCancel context.CancelFunc
+	transport    string
+}
+
+func NewReceiver(
+	_ context.Context,
+	config *Config,
+	logger *zap.Logger,
+	nextConsumer consumer.MetricsConsumerOld,
+) (component.MetricsReceiver, error) {
+	err := config.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	parsed, err := url.Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("could not determine receiver transport: %w", err)
+	}
+
+	receiver := Receiver{
+		config:       config,
+		nextConsumer: nextConsumer,
+		logger:       logger,
+		transport:    parsed.Scheme,
+	}
+
+	return &receiver, nil
+}
+
+func (r *Receiver) Start(ctx context.Context, host component.Host) error {
+	var err error
+	r.client, err = NewDockerClient(r.config, r.logger)
+	if err != nil {
+		return err
+	}
+
+	r.obsCtx = obsreport.ReceiverContext(ctx, typeStr, r.transport, r.config.Name())
+
+	runnableCtx, cancel := context.WithCancel(ctx)
+
+	runnable := newRunnable(runnableCtx, cancel, r)
+	r.runner = interval.NewRunner(r.config.CollectionInterval, runnable)
+	r.runnerCancel = cancel
+
+	go func() {
+		if err := r.runner.Start(); err != nil {
+			host.ReportFatalError(err)
+		}
+	}()
+
+	return nil
+}
+
+func (r *Receiver) Shutdown(ctx context.Context) error {
+	r.runnerCancel()
+	r.runner.Stop()
+	return nil
+}
+
+type Runnable struct {
+	ctx               context.Context
+	cancel            func()
+	receiver          *Receiver
+	successfullySetup bool
+}
+
+var _ interval.Runnable = (*Runnable)(nil)
```

Comment:
I feel like it might be time to move the runnable stuff out of the redis receiver directory and to a location more appropriate for general use. :-)

Reply:
May be worth syncing with the team on the best way to handle 'scrapers' since this is such a common requirement.

Reply:
I agree and intended to do so in a subsequent PR so that we can better standardize the interval api.